<launch>
  <!-- Static Transforms -->
 <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_odom" args="0 0 0 0 0 0 /map /odom_combined" />   
  <node pkg="tf2_ros" type="static_transform_publisher" name="base_link_to_laser" args="0.06 0 0.08 0 0 0 /base_link /laser" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="base_footprint_to_base_link" args="0.0 0 0.08 0 0 0 /base_footprint /base_link" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="base_link_to_imu" args="0.0 0 0.08 0 0 0 /base_link /imu" />
  <!-- <node pkg="tf2_ros" type="static_transform_publisher" name="map_to_ekf_odom" args="0 0 0 0 0 0 /map /ekf_odom" /> -->


 <node pkg="rosserial_python" type="serial_node.py" name="serial_node">
    <param name="port" value="/dev/ttyACM0"/>
    <param name="baud" value="57600"/>
  </node> 
   
   <!-- <param name="robot_description" textfile="$(find navigation)/urdf/final.urdf" />     -->

  <node pkg="navigation" type="ekf_odom_pub" name="ekf_odom_pub">
       <remap from="odom_data_quat" to="odom"/> 

  </node>

    <!-- <node pkg="navigation" type="odom_to_base" name="odom_to_base">
  </node>    -->

<node pkg="navigation" type="imu_node.py" name="imu_node" output="screen">
    <param name="i2c_device" value="/dev/i2c-1" />
    <param name="device_address" value="0x68" />
    <param name="frame_id" type="string" value="imu"/>   
       </node>

   
  <!-- Start robot_pose_ekf node -->
  
  <!-- Start robot_pose_ekf node -->
  <!-- Extended Kalman Filter from robot_pose_ekf Node-->
  <!-- Subscribe: /odom, /imu_data, /vo -->
  <!-- Publish: /robot_pose_ekf/odom_combined -->
   
   <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
    <param name="output_frame" value="odom_combined"/>
      <remap from="imu_data" to="imu/data" />
    <param name="base_footprint_frame" value="base_footprint"/>
    <param name="base_frame" value="base_link"/>
    <param name="world_frame" value="odom"/>
    <rosparam command="load" file="$(find navigation)/config/robot_pose_ekf.yaml"/>
  </node> 

  <!-- <arg name="rviz_config_file" default="$(find navigation)/config/robot.rviz"/>
  <arg name="model" />
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rviz_config_file)"/>  -->
     <node pkg="localization" type="rviz_click_to_2d" name="rviz_click_to_2d" />


<node name="rplidarNode" pkg="rplidar_ros" type="rplidarNode" output="screen">
    <param name="serial_port" type="string" value="/dev/ttyUSB0"/>
    <param name="serial_baudrate" type="int" value="115200"/>
    <param name="frame_id" type="string" value="laser"/>
    <param name="inverted" type="bool" value="false"/>
    <param name="angle_compensate" type="bool" value="true"/>
  </node>

  <!-- Wheel Encoder Tick Publisher and Base Controller Using Arduino -->
  

  <!-- Wheel Odometry Publisher -->
  <!-- <node pkg="navigation" type="map_to_odo" name="map_to_odo">
  </node> -->

    <arg name="map_file" default="$(find navigation)/maps/map.yaml"/> 
  
  <!-- Map Server -->
    <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)" />  
   
    
    <!-- Start AMCL node -->
  <arg name="scan_topic"     default="scan"/>
  <arg name="initial_pose_x" default="0.0"/>
  <arg name="initial_pose_y" default="0.0"/>
  <arg name="initial_pose_a" default="0.0"/>

  <!-- AMCL -->
    <arg name="use_map_topic" default="true"/>
  <node pkg="amcl" type="amcl" name="amcl">
    <param name="use_map_topic" value="$(arg use_map_topic)"/>
  <!-- Publish scans from best pose at a max of 10 Hz -->
  <param name="odom_model_type" value="omni"/>
  <param name="odom_alpha5" value="0.1"/>
  <param name="gui_publish_rate" value="10.0"/>
  <param name="laser_max_beams" value="30"/>
  <param name="min_particles" value="500"/>
  <param name="max_particles" value="5000"/>
  <param name="kld_err" value="0.05"/>
  <param name="kld_z" value="0.99"/>
  <param name="odom_alpha1" value="0.2"/>
  <param name="odom_alpha2" value="0.2"/>
  <!-- translation std dev, m -->
  <param name="odom_alpha3" value="0.8"/>
  <param name="odom_alpha4" value="0.2"/>
  <param name="laser_z_hit" value="0.5"/>
  <param name="laser_z_short" value="0.05"/>
  <param name="laser_z_max" value="0.05"/>
  <param name="laser_z_rand" value="0.5"/>
  <param name="laser_sigma_hit" value="0.2"/>
  <param name="laser_lambda_short" value="0.1"/>
  <param name="laser_model_type" value="likelihood_field"/>
  <!-- <param name="laser_model_type" value="beam"/> -->
  <param name="laser_likelihood_max_dist" value="2.0"/>
  <param name="update_min_d" value="0.2"/>
  <param name="update_min_a" value="0.5"/>
  <param name="odom_frame_id" value="odom_combined"/>
  <param name="resample_interval" value="1"/>
  <param name="transform_tolerance" value="0.1"/>
  <param name="recovery_alpha_slow" value="0.0"/>
  <param name="recovery_alpha_fast" value="0.0"/>
  <param name="base_frame_id"   value="base_link"/>

</node>
  <!-- Move Base Node with Costmap Parameters -->
  
   <arg name="odom_frame_id" default="odom_combined"/>
  <arg name="cmd_vel_topic" default="/cmd_vel" />
  <arg name="odom_topic" default="/odom" />
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">

  <param name="footprint_padding" value="0.01" />
    <param name="controller_frequency" value="10.0" />
    <param name="controller_patience" value="3.0" />

    <param name="oscillation_timeout" value="30.0" />
    <param name="oscillation_distance" value="0.5" />

    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />

    <rosparam file="$(find navigation)/param/costmap_common_params_burger.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find navigation)/param/costmap_common_params_burger.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find navigation)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find navigation)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find navigation)/param/move_base_params.yaml" command="load" /> 
    <rosparam file="$(find navigation)/param/dwa_local_planner_params_burger.yaml" command="load" />
    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <remap from="map" to="map"/>
  </node>







  <!-- <arg name="base_frame"  default="base_link"/>
  <arg name="odom_frame"  default="odom_combined"/>
  <arg name="map_frame"  default="map"/> -->
  
<!-- <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen">
   <param name="base_frame" value="$(arg base_frame)"/>
    <param name="odom_frame" value="$(arg odom_frame)"/>
    <param name="map_frame"  value="$(arg map_frame)"/>

    
  </node> -->
 

</launch>